janus_core/calculations/eos.py

```diff
@@ -0,0 +1,137 @@
+    n_lattice: int = 7,
+    eos_type: EoSNames = "birchmurnaghan",
+    minimize: bool = True,
+    minimize_kwargs: Optional[dict[str, Any]] = None,
+    write_results: bool = True,
+    file_prefix: Optional[PathLike] = None,
+    log_kwargs: Optional[dict[str, Any]] = None,
+) -> EoSResults:
+    """
+    Calculate equation of state.
+
+    Parameters
+    ----------
+    struct : Atoms
+        Structure.
+    struct_name : Optional[str]
+        Name of structure. Default is None.
+    min_lattice : float
+        Minimum lattice constant scale factor. Default is 0.95.
+    max_lattice : float
+        Maximum lattice constant scale factor. Default is 1.05.
+    n_lattice : int
+        Number of lattice constants to use. Default is 7.
+    eos_type : EoSNames
+        Type of fit for equation of state. Default is "birchmurnaghan".
+    minimize : bool
+        Whether to optimize geometry. Default is True.
+    minimize_kwargs : Optional[dict[str, Any]]
+        Keyword arguments to pass to optimize. Default is None.
+    write_results : bool
+        True to write out results of equation of state calculations. Default is True.
+    file_prefix : Optional[PathLike]
+        Prefix for output filenames. Default is inferred from structure name, or
+        chemical formula of the structure.
+    log_kwargs : Optional[dict[str, Any]]
+        Keyword arguments to pass to `config_logger`. Default is {}.
+
+    Returns
+    -------
+    EoSResults
+        Dictionary containing equation of state ASE object, and fitted minimum
+        bulk_modulus, volume, and energy.
+    """
+    [minimize_kwargs, log_kwargs] = none_to_dict([minimize_kwargs, log_kwargs])
+    log_kwargs.setdefault("name", __name__)
+    logger = config_logger(**log_kwargs)
+
+    struct_name = struct_name if struct_name else struct.get_chemical_formula()
+    file_prefix = file_prefix if file_prefix else struct_name
+
+    if not struct.calc:
+        raise ValueError("Please attach a calculator to `struct`.")
+
+    # Ensure lattice constants span correct range
+    if n_lattice <= 1:
+        raise ValueError("`n_lattice` must be greater than 1.")
+    if not 0 < min_lattice < 1:
+        raise ValueError("`min_lattice` must be between 0 and 1.")
+    if max_lattice <= 1:
+        raise ValueError("`max_lattice` must be greater than 1.")
+
+    if minimize:
+        if logger:
+            minimize_kwargs["log_kwargs"] = {
+                "filename": log_kwargs["filename"],
+                "name": logger.name,
+                "filemode": "a",
+            }
+        optimize(struct, **minimize_kwargs)
+
+    if logger:
+        logger.info("Starting calculations for configurations")
+
+    cell = struct.get_cell()
+
+    lattice_scalars = np.linspace(min_lattice, max_lattice, n_lattice) ** (1 / 3)
+    volumes = []
+    energies = []
+    for lattice_scalar in lattice_scalars:
+        struct.set_cell(cell * lattice_scalar, scale_atoms=True)
```

Comment:
```suggestion
        struct.set_cell(cell * lattice_scalar, scale_atoms=True)
        if minimize:
           optimize(struct, constant_volume=True, **minmize_kwargs)
```

Reply:
what do you mean?

Reply:
I thought you don't always want to minimise the structure for all the lattice constants, but you may still want to minimize before starting at all?

I added a separate option, so you can do both or only before starting, but I didn't make sure the volume is conserved yet

Reply:
yap is fine, we also need to save initial optimized cell and coordinates and use that before the optimisation, calculation of the energy in the loop.